src/librbd/cache/rwl/LogOperation.h

```diff
@@ -0,0 +1,203 @@
+
+namespace librbd {
+namespace cache {
+namespace rwl {
+struct WriteBufferAllocation;
+
+class WriteLogOperationSet;
+
+class WriteLogOperation;
+
+class GenericWriteLogOperation;
+
+class SyncPointLogOperation;
+
+class GenericLogOperation;
+
+using GenericLogOperationSharedPtr = std::shared_ptr<GenericLogOperation>;
+
+using GenericLogOperationsVector = std::vector<GenericLogOperationSharedPtr>;
+
+class GenericLogOperation {
+protected:
+  PerfCounters *m_perfcounter = nullptr;
+public:
+  utime_t dispatch_time;         // When op created
+  utime_t buf_persist_time;      // When buffer persist begins
+  utime_t buf_persist_comp_time; // When buffer persist completes
+  utime_t log_append_time;       // When log append begins
+  utime_t log_append_comp_time;  // When log append completes
+  GenericLogOperation(const utime_t dispatch_time, PerfCounters *perfcounter);
+  virtual ~GenericLogOperation() { };
+  GenericLogOperation(const GenericLogOperation&) = delete;
+  GenericLogOperation &operator=(const GenericLogOperation&) = delete;
+  virtual std::ostream &format(std::ostream &os) const;
+  friend std::ostream &operator<<(std::ostream &os,
+                                  const GenericLogOperation &op);
+  virtual const std::shared_ptr<GenericLogEntry> get_log_entry() = 0;
+  virtual const std::shared_ptr<WriteLogEntry> get_write_log_entry() {
+    return nullptr;
+  }
+  virtual void appending() = 0;
+  virtual void complete(int r) = 0;
+  virtual bool is_write() {
+    return false;
+  }
+  virtual bool is_sync_point() {
+    return false;
+  }
+  virtual bool is_discard() {
+    return false;
+  }
+  virtual bool is_writesame() {
+    return false;
+  }
+  virtual bool is_writing_op() {
+    return false;
+  }
+  virtual GenericWriteLogOperation *get_gen_write_op() {
+    return nullptr;
+  };
+  virtual WriteLogOperation *get_write_op() {
+    return nullptr;
+  };
+  virtual bool reserved_allocated() {
+    return (is_write() || is_writesame());
+  }
+};
+
+class SyncPointLogOperation : public GenericLogOperation {
+private:
+  CephContext *m_cct;
+  ceph::mutex &m_lock;
+  std::vector<Context*> append_sync_point();
+  void clear_earlier_sync_point();
+  std::vector<Context*> swap_on_sync_point_persisted();
+public:
+  std::shared_ptr<SyncPoint> sync_point;
+  SyncPointLogOperation(ceph::mutex &lock,
+                        std::shared_ptr<SyncPoint> sync_point,
+                        const utime_t dispatch_time,
+                        PerfCounters *perfcounter,
+                        CephContext *cct);
+  ~SyncPointLogOperation() override;
+  SyncPointLogOperation(const SyncPointLogOperation&) = delete;
+  SyncPointLogOperation &operator=(const SyncPointLogOperation&) = delete;
+  std::ostream &format(std::ostream &os) const;
+  friend std::ostream &operator<<(std::ostream &os,
+                                  const SyncPointLogOperation &op);
+  const std::shared_ptr<GenericLogEntry> get_log_entry() {
```

Comment:
Nit: this applies for remainder of methods here (and the virtual methods in other derived classes), but make sure you append `override` suffix on virtual method overrides

Reply:
Set it override.